File: internal/ats/lever.go
package ats

import (
	"fmt"
	"time"

	"github.com/Vaivaswat2244/job-tracker/internal/db"
	"github.com/Vaivaswat2244/job-tracker/internal/httpx"
	"github.com/Vaivaswat2244/job-tracker/internal/textutil"
)

// Lever reads the postings API:
//
//	GET https://api.lever.co/v0/postings/{slug}?mode=json
//
// It returns a bare array. `createdAt` is epoch MILLISECONDS, not seconds —
// divide before converting or every posting lands in 1970.
type Lever struct{}

const leverAPI = "https://api.lever.co/v0/postings/%s?mode=json"

func (Lever) Name() string { return "lever" }

func (Lever) BoardURL(slug string) string { return "https://jobs.lever.co/" + slug }

func epochMS(value any) *string {
	ms, ok := textutil.AsFloat(value)
	if !ok {
		return nil
	}
	sec, frac := int64(ms/1000), ms-float64(int64(ms/1000))*1000
	t := time.Unix(sec, int64(frac)*int64(time.Millisecond)).UTC()
	s := t.Format(db.ISO8601)
	return &s
}

func (l Lever) Parse(payload any) []NormalizedJob {
	jobs := []NormalizedJob{}
	items, _ := payload.([]any)
	for _, raw := range items {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		externalID, hasID := textutil.AsString(item["id"])
		title, hasTitle := textutil.AsString(item["text"])
		if !hasID || !hasTitle {
			continue
		}
		url, _ := textutil.AsString(item["hostedUrl"])

		// descriptionPlain is authoritative when present; the HTML field is the
		// fallback and needs tags stripped.
		jd := ""
		if body, ok := textutil.AsString(item["descriptionPlain"]); ok {
			jd = textutil.Clean(body)
		} else {
			html, _ := textutil.AsString(item["description"])
			jd = textutil.StripHTML(html)
		}

		jobs = append(jobs, NormalizedJob{
			ExternalID:     externalID,
			Title:          title,
			URL:            url,
			Source:         l.Name(),
			PostedAt:       epochMS(item["createdAt"]),
			Location:       optString(textutil.Dig(item, "categories", "location")),
			EmploymentType: optString(textutil.Dig(item, "categories", "commitment")),
			JDText:         jd,
			Raw:            item,
		})
	}
	return jobs
}

func (l Lever) Fetch(slug, etag, lastModified string) FetchResult {
	resp := httpx.Get(fmt.Sprintf(leverAPI, slug), httpx.Options{
		ETag: etag, LastModified: lastModified,
	})
	return finish(l, resp, etag, lastModified, func(payload any) (any, string) {
		if _, ok := payload.([]any); !ok {
			return nil, "expected a JSON array of postings"
		}
		return payload, ""
	})
}

File: tracker/ats/lever.py
"""Lever postings API.

GET https://api.lever.co/v0/postings/{slug}?mode=json
Returns a bare array. `createdAt` is epoch MILLISECONDS, not seconds — divide
before converting or every posting lands in 1970.
"""
from datetime import datetime, timezone

from .. import http
from ..textutil import as_str, clean, dig, strip_html
from . import FetchResult, NormalizedJob

API = "https://api.lever.co/v0/postings/{slug}?mode=json"
SOURCE = "lever"


def board_url(slug: str) -> str:
    return f"https://jobs.lever.co/{slug}"


def _epoch_ms(value) -> str | None:
    try:
        ms = float(value)
    except (TypeError, ValueError):
        return None
    return datetime.fromtimestamp(ms / 1000.0, tz=timezone.utc).isoformat(timespec="seconds")


def parse(payload) -> list[NormalizedJob]:
    jobs = []
    for item in payload or []:
        if not isinstance(item, dict):
            continue
        external_id = as_str(item.get("id"))
        title = as_str(item.get("text"))
        if not external_id or not title:
            continue
        body = as_str(item.get("descriptionPlain"))
        jobs.append(
            NormalizedJob(
                external_id=external_id,
                title=title,
                url=as_str(item.get("hostedUrl")) or "",
                source=SOURCE,
                posted_at=_epoch_ms(item.get("createdAt")),
                location=as_str(dig(item, "categories", "location")),
                employment_type=as_str(dig(item, "categories", "commitment")),
                jd_text=clean(body) if body else strip_html(item.get("description")),
                raw=item,
            )
        )
    return jobs


def fetch(slug: str, *, etag=None, last_modified=None) -> FetchResult:
    resp = http.get(API.format(slug=slug), etag=etag, last_modified=last_modified)
    if resp.not_modified:
        return FetchResult(jobs=None, status=304, not_modified=True,
                           etag=etag, last_modified=last_modified)
    if not resp.ok:
        return FetchResult(jobs=None, status=resp.status,
                           error=resp.error or f"HTTP {resp.status}")
    payload = resp.json()
    if not isinstance(payload, list):
        return FetchResult(jobs=None, status=resp.status,
                           error="expected a JSON array of postings")
    return FetchResult(jobs=parse(payload), status=resp.status,
                       etag=resp.etag, last_modified=resp.last_modified)
